xds/src/test/java/io/grpc/xds/XdsLrsClientTest.java

```diff
@@ -0,0 +1,379 @@
+import java.util.LinkedList;
+import java.util.concurrent.TimeUnit;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+/** Unit tests for {@link XdsLrsClient}. */
+public class XdsLrsClientTest {
+
+  private static final String SERVICE_AUTHORITY = "api.google.com";
+  private static final FakeClock.TaskFilter LOAD_REPORTING_TASK_FILTER =
+      new FakeClock.TaskFilter() {
+        @Override
+        public boolean shouldAccept(Runnable command) {
+          return command.toString().contains(XdsLrsClient.LoadReportingTask.class.getSimpleName());
+        }
+      };
+  private static final FakeClock.TaskFilter LRS_RPC_RETRY_TASK_FILTER =
+      new FakeClock.TaskFilter() {
+        @Override
+        public boolean shouldAccept(Runnable command) {
+          return command.toString().contains(XdsLrsClient.LrsRpcRetryTask.class.getSimpleName());
+        }
+      };
+  @Rule
+  public final GrpcCleanupRule cleanupRule = new GrpcCleanupRule();
+  private final SynchronizationContext syncContext = new SynchronizationContext(
+      new Thread.UncaughtExceptionHandler() {
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+          throw new AssertionError(e);
+        }
+      });
+  private final ArrayList<String> logs = new ArrayList<>();
+  private final ChannelLogger channelLogger = new ChannelLogger() {
+    @Override
+    public void log(ChannelLogLevel level, String msg) {
+      logs.add(level + ": " + msg);
+    }
+
+    @Override
+    public void log(ChannelLogLevel level, String template, Object... args) {
+      log(level, MessageFormat.format(template, args));
+    }
+  };
+  private LoadReportingServiceGrpc.LoadReportingServiceImplBase mockLoadReportingService;
+  private FakeClock fakeClock = new FakeClock();
+  private final LinkedList<StreamObserver<LoadStatsRequest>> lrsRequestObservers =
+      new LinkedList<>();
+  @Captor
+  private ArgumentCaptor<StreamObserver<LoadStatsResponse>> lrsResponseObserverCaptor;
+
+  @Mock
+  private Helper helper;
+  @Mock
+  private BackoffPolicy.Provider backoffPolicyProvider;
+  private static final LoadStatsRequest EXPECTED_INITIAL_REQ = LoadStatsRequest.newBuilder()
+      .setNode(Node.newBuilder()
+          .setMetadata(Struct.newBuilder()
+              .putFields(
+                  XdsLrsClient.TRAFFICDIRECTOR_HOSTNAME_FIELD,
+                  Value.newBuilder().setStringValue(SERVICE_AUTHORITY).build())))
+      .build();
+  @Mock
+  private BackoffPolicy backoffPolicy1;
+  private XdsLrsClient lrsClient;
+  @Mock
+  private BackoffPolicy backoffPolicy2;
+
+  private static ClusterStats buildEmptyClusterStats(long loadReportIntervalNanos) {
```

Comment:
@zhangkun83 I don't think we need to have tests that involve reporting actual load data (by crafting data in `XdsLoadReportStore`) as we already have tests related to `generateLoadReport` in `XdsLoadReportStore`'s own unit tests. Here, I believe testing interactions between the balancer and `XdsLrsClient` would suffice. Do you think so?

Reply:
Because `XdsLoadReportStore` is not an interface, the only way to verify it's actually called for each report is to craft data so that each report is distinguishable. We don't want to craft the data variation in each of the tests (it's fine to use empty reports for them), but we should have one test case that specifically does that.

Reply:
ACK.